File: xgraphics/src/js/x/graphics/primitive.js
import Obj from './obj.js';
import Vertices from './vertices.js';
import Buffer from './buffer.js';
import Indices from './indices.js';
import Vertex from './vertex.js';
import Matrix from './matrix.js';

export default class Primitive extends Obj
{
    constructor(name, level, program)
    {
        super(name, level);

        this.__vertices = new Vertices();
        this.__buffer = new Buffer();
        this.__program = program;
        this.__indices = new Indices();
    }

    push(o)
    {
        this.__vertices.push(o);
    }

    load(application)
    {
        if(!this.__load)
        {
            if(!this.__buffer.size)
            {
                this.__vertices.serialize(this.__buffer);
            }

            this.__program.load(application);
            this.__indices.load(application);
            this.__vertices.load(application);
            this.__buffer.load(application);
            this.__load = true;
        }
    }

    unload(application)
    {
        if(this.__load)
        {
            this.__program.unload(application);
            this.__indices.unload(application);
            this.__vertices.unload(application);
            this.__buffer.unload(application);
            this.__load = false;
        }
    }

    bind(application, name, value, model, level)
    {
        const gl = application.gl;
        if(name === "mvp")
        {
            const camera = application.camera;
            const projection = application.projection;
            // console.log(model)
            let mvp = Matrix.multiply(Matrix.multiply(model, camera.matrix), projection.matrix);
            gl.uniformMatrix4fv(value, false, new Float32Array(mvp));
            return true;
        }
        else if(name === "cameraPos")
        {
            const camera = application.camera;
            gl.uniform4fv(value, camera.location);
            return true;
        }
        else if(name === "lightColor")
        {
            const light = application.light;
            gl.uniform4fv(value, light.color);
            return true;
        }
        else if(name === "lightPos")
        {
            const light = application.light;
            gl.uniform4fv(value, light.location);
            return true;
        }
        else if(name === "ambientStrength")
        {
            const light = application.light;
            gl.uniform1f(value, light.ambient);
            return true;
        }
        else
        {
            return false;
        }
    }

    render(application, model, level)
    {
        if(this.__level === level)
        {
            const gl = application.gl;

            gl.useProgram(this.__program.id);
            this.__program.uniforms.forEach((value, name) => {
                this.bind(application, name, value, model, level);
            });
            this.__program.attributes.forEach((value, name) => {
                if(value >= 0)
                {
                    gl.enableVertexAttribArray(value);
                    this.buffer.bind(application);
                    if(name === "vertexPosition")
                    {
                        gl.vertexAttribPointer(value, 4, gl.FLOAT, false, Vertex.stride, 0);
                    }
                    else if(name === "vertexNormal")
                    {
                        gl.vertexAttribPointer(value, 4, gl.FLOAT, false, Vertex.stride, 4 * 4);
                    }
                    else if(name === "vertexUV")
                    {
                        gl.vertexAttribPointer(value, 4, gl.FLOAT, false, Vertex.stride, 12 * 4);
                    }
                }
            });
            this.buffer.bind(application);
            if(this.__indices.size)
            {
                this.__indices.bind(application);
                // TODO: REMOVE TRIANGLES
                gl.drawElements(gl.TRIANGLES, this.__indices.size, gl.UNSIGNED_SHORT, 0);
            }
            else
            {
                // TODO: REMOVE TRIANGLES
                gl.drawArrays(gl.TRIANGLES, 0, this.__vertices.size);
            }
        }
    }

    get name() { return this.__name; }
    get vertices() { return this.__vertices; }
    get buffer(){ return this.__buffer; }
    get program() { return this.__program; }

    set mode(v){ this.__vertices.mode = v; }
    set target(v){ this.__buffer.target = v; }
    set usage(v){ this.__buffer.usage = v; }
    set indices(v){ this.__indices = new Indices(v); }
    set status(v){ this.__status = v; }
    set program(v){ this.__program = v; }
}

File: 2020/old/discomplexity/x/src/x/graphics/primitive.cc
#include "primitive.hh"
#include "application.hh"

#include "opengl.hh"

namespace x
{
    namespace graphics
    {
        primitive::primitive(const x::string & name)
        :   x::graphics::object(name),
            __vertices(),
            __buffer(),
            __program(nullptr),
            __indices()
        {

        }
        primitive::primitive(x::string && name)
        :   x::graphics::object(x::move(name)),
            __vertices(),
            __buffer(),
            __program(nullptr),
            __indices()
        {

        }

        primitive::primitive()
        :   x::graphics::object(),
            __vertices(),
            __buffer(),
            __program(nullptr),
            __indices()
        {
        }

        primitive::primitive(const primitive & o)
        :   x::graphics::object(o),
            __vertices(o.__vertices),
            __buffer(o.__buffer),
            __program(o.__program),
            __indices(o.__indices)
        {
        }

        primitive::primitive(primitive && o) noexcept
        :   x::graphics::object(x::move(o)),
            __vertices(x::move(o.__vertices)),
            __buffer(x::move(o.__buffer)),
            __program(x::move(o.__program)),
            __indices(x::move(o.__indices))
        {
            o.__program = nullptr;
        }

        primitive::~primitive()
        {
        }

        primitive & primitive::operator=(const primitive & o)
        {
            if(&o != this)
            {
                x::graphics::object::operator=(o);
                __vertices = o.__vertices;
                __buffer = o.__buffer;
                __program = o.__program;
                __indices = o.__indices;
            }
            return *this;
        }

        primitive & primitive::operator=(primitive && o) noexcept
        {
            if(&o != this)
            {
                x::graphics::object::operator=(x::move(o));
                __vertices = x::move(o.__vertices);
                __buffer = x::move(o.__buffer);
                __program = x::move(o.__program);
                __indices = x::move(o.__indices);

                o.__program = nullptr;
            }
            return *this;
        }

        void primitive::load()
        {
            x::graphics::application & application = x::graphics::application::get();
            // x::console::out << "load primitive => " << x::endl;
            application.load(*this);
        }

        void primitive::unload()
        {
            x::graphics::application & application = x::graphics::application::get();

            application.unload(*this);
        }

        void primitive::render(const x::graphics::matrix & model, x::uint32 level)
        {
            if(__status != 0)
            {
                if(level == __level)
                {
                    x::graphics::application & application = x::graphics::application::get();
                    const x::graphics::camera & camera = application.camera();
                    const x::graphics::light & light = application.light();
                    const x::graphics::projection & projection = application.projection();

                    x::graphics::matrix mvp = model * camera.matrix() * projection.matrix();

                    application.bind(__vertices);
                    if(__program)
                    {
                        application.bind(*__program);
                        // TODO REMOVE THIS
                        application.bind(*__program, "mvp", mvp);
                        application.bind(*__program, "cameraPos", camera.position());
                        application.bind(*__program, "lightPos", light.position());
                        application.bind(*__program, "lightColor", light.color());
                        application.bind(*__program, "ambientStrength", light.ambient());
                    }
                    else
                    {
                        x::console::out << "program not exist" << x::endl;
                    }


                    application.bind(__buffer);
                    application.bind(0, __buffer, x::graphics::vertex::stride, 0);
                    application.bind(1, __buffer, x::graphics::vertex::stride, 4 * sizeof(GLfloat));
                    application.bind(2, __buffer, x::graphics::vertex::stride, 8 * sizeof(GLfloat));
                    application.draw(__vertices, __indices);
                    application.unbind(2);
                    application.unbind(1);
                    application.unbind(0);
                }
            }
        }
    }
}
